java-checks/src/main/java/org/sonar/java/checks/ConstantsShouldBeStaticFinalCheck.java

```diff
@@ -96,7 +96,14 @@ private static boolean hasConstantInitializer(VariableTree variableTree) {
       if (init.is(Tree.Kind.NEW_ARRAY)) {
         return false;
       }
-      return !containsChildrenOfKind((JavaTree) init, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+      return !containsChildMatchingPredicate((JavaTree) init, tree -> {
+          if (tree.is(Tree.Kind.IDENTIFIER)) {
```

Comment:
Would be nice to extract it to a separate method. And combine these 2 conditions using `Predicate.or()` if possible.